D8CORE-2106

imonroe opened the pull request:
# READY FOR REVIEW

# Summary
- The link field widget needs to be able to support cardinality in order to allow more than one link to be used in the CTA List paragraph type.

# Review By (Date)
- ASAP

# Urgency
- These changes unblocks D8CORE-1656

# Steps to Test

1. Build a stack with this branch of the react_paragraphs module.
2. Add a paragraph with a link field (or create a new paragraph type with a link field)
3. Add a link.
4. Click the "Add Another Link" button
5. Notice it adds an additional link to the list.  
6. Add an additional link.  Save the paragraph/page.
7. Edit the paragraph.  Click on the "Remove" button next to the second link's `Link Text` field.
8. Notice it removes the additional link.
9. Save the paragraph.  Make sure it displays only your first link correctly. 

# Affected Projects or Products
- SOE custom paragraph types.

# Associated Issues and/or People
- D8CORE-1656

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

imonroe committed:
D8CORE-2016: Got link field widget to display multiple cardinality, and update the existing values when they are edited.

imonroe committed:
D8CORE-2016: Added Add-another-link button, corrected autocomplete functionality

imonroe committed:
D8CORE-2016: Cleanup debugging

imonroe committed:
D8CORE-2016: Cleanup unused variable.

imonroe committed:
D8CORE-2106: Added logic to deal with empty initial fields

pookmish commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -1,20 +1,44 @@
-    uri: defaultValue && defaultValue.length ? defaultValue[0].uri : '',
-    title: defaultValue && defaultValue.length ? defaultValue[0].title : ''
-  };
+  function emptyLinkValue() {
```
this isn't the consistent way to define a function. and because this doesn't do anything, it could just be a `const` at the top of the component.

imonroe replied:
I tried it as a `const` and it didn't work.  When I changed it to a function, it worked fine.  I'm not sure why that's true, but I'd love to know.

imonroe replied:
This is now fixed.

pookmish commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -1,20 +1,44 @@
+    };
+  }
+
+  const alterValues = (props) => {
```
`props` is a common variable in stateful components. the variable here should be named something different.

imonroe replied:
fixed.

pookmish commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -1,20 +1,44 @@
+  }
+
+  const alterValues = (props) => {
+    const newState = Array.from(fieldValues);
```
```suggestion
    const newState = [...fieldValues];
```

pookmish commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -1,20 +1,44 @@
+    const newState = Array.from(fieldValues);
+    newState[props.delta].title = props.title;
+    newState[props.delta].uri = props.uri;
+    setValues(newState);
+    onFieldChange(fieldValues);
```
why do `setValues` and `onFieldChange` use different variables?

imonroe replied:
So, in this case, `setValues` updates the local state, which was set up via the `useState` hook in line 19.  `fieldValues` is the current local state.  `onFieldChange` is the function that is passed in via the constructor, which handles the state one layer up, in the `ParagraphForm.js`, and does the work of setting the field in Drupal, right?

So I created the local state so I could change it, since the widget is declared fully functionally, rather than object oriented, otherwise I couldn't change it inside the widget.  When the local state is altered, I feed that back up into the `onFieldChange` function so that it'll persist outside of the widget.

imonroe replied:
This is now fixed to use only the onFieldChange function.

pookmish commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -118,48 +142,64 @@ export const LinkWidget = ({fieldId, defaultValue, onFieldChange, settings}) =>
+  /**
+   * To support cardinality, we will need an "Add another" button.
+   */
+  const addAnotherButton = (cardinality) => {
```
theres no need for passing `cardinality` here. the function will already have access to `settings.cardinality`

imonroe replied:
fixed

pookmish commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -118,48 +142,64 @@ export const LinkWidget = ({fieldId, defaultValue, onFieldChange, settings}) =>
+   * To support cardinality, we will need an "Add another" button.
+   */
+  const addAnotherButton = (cardinality) => {
+    if ((cardinality == -1) || (cardinality < fieldValues.length )) {
```
```suggestion
    if ((cardinality == -1) || (cardinality > fieldValues.length )) {
```

pookmish reviewed (changes requested):
This looks like a good approach. its very much like I would've done as well. just some adjustments and small tweaks necessary.

imonroe committed:
D8CORE-2106: adding fixes from feedback

imonroe committed:
Apply suggestions from code review

Co-authored-by: pookmish <pookmish@stanford.edu>

imonroe committed:
D8CORE-2106: recompiled assets with changes

imonroe committed:
D8CORE-2106: Added button to remove extra links, cleanup

pookmish commented:
I'm getting an error in the console and all the react widget disappears. This is on a new piece of content with no default values. I believe its because of [default value being null](https://github.com/SU-SWS/react_paragraphs/pull/48/files#diff-27eaa7ec2543dbf24b08100c65c0e56dR12)
![image](https://user-images.githubusercontent.com/7185045/84063767-6b521d00-a976-11ea-8aa2-aff08efb49cc.png)

imonroe committed:
D8CORE-2106: Changed when the `fieldValues` gets initialized to get around problems with empty fields.

imonroe commented:
Ok, I've changed the order of operations a little bit, so it doesn't initiate `fieldValues` until after it checks to make sure to that there's at least one element in the array.  That fixes it for me.

pookmish commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -118,48 +135,95 @@ export const LinkWidget = ({fieldId, defaultValue, onFieldChange, settings}) =>
+   * If we add more links, we need a way of removing them, but never remove link 0.
+   */
+  const removeLinkButton = (delta) => {
+    if (delta > 0){
```
It would be nice to be able to delete the first link as well.

imonroe replied:
Fixed.  In order to prevent re-draws with an empty array, I have it add a new, empty link if the count of the array reaches zero.  That should prevent render bugs if all the links are deleted.

pookmish commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -118,48 +135,95 @@ export const LinkWidget = ({fieldId, defaultValue, onFieldChange, settings}) =>
+   * Handler for removing a link.
+   */
+  const removeLink = (delta) => {
+    if (fieldValues[delta] !== undefined){
```
i think its best to use `typeof fieldValues[delta] !== 'undefined'`

imonroe replied:
Fixed.

pookmish commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -118,48 +135,95 @@ export const LinkWidget = ({fieldId, defaultValue, onFieldChange, settings}) =>
+   */
+  const removeLink = (delta) => {
+    if (fieldValues[delta] !== undefined){
+      let removed = fieldValues.splice(delta, 1);
```
You don't need to set the variable if you're not using it anywhere

imonroe replied:
fixed

pookmish reviewed (commented):
This works great. only the one thing about deleting the first one. it shouldn't really change the delete logic at all.

Can you also pass off a screenshot to Ishi or Kerri for a quick review

imonroe committed:
D8CORE-2106: Added the ability to remove the first link from the field, plus a little cleanup.

imonroe commented:
I've slacked a screenshot of the interface over to Ishi and Kerri for any feedback they may have.

imonroe committed:
D8CORE-2106: Updated button styles to match Drupal admin, removed unneeded dependency

imonroe commented:
Ishi asked that the button styles be changed to match the Drupal admin buttons, so I changed them from `<Button>` to `<input>`  elements, and removed the material-ui button dependency.

pookmish commented:
There are a couple new issues now. 

1. a user can delete the first link even if its the only one. that should be prevented
2. the add button now closes the dialog instead of adding another row and staying in the dialog. same with the remove button.

be sure to review any button in drupal. they act differently than they do in the dev browser

imonroe commented:
So, if a user deletes the first link, it replaces it with an empty one.  Shall I make it so that there is no "Remove" button if there is less that two links, but if there are two or more, then each link gets a "Remove" button?

Also, I am not able to reproduce the behavior with the buttons now closing the dialog.  They work as expected for me, within a full Drupal install, using both Chrome and Firefox.  Did you clear your cache when testing?

imonroe committed:
D8CORE-2106: Remove button now only appears if there is more than one link in the field

imonroe commented:
Ok, I set it so the remove buttons only show up if there's more than one link.

pookmish commented:
One solution to the buttons is to use a button element like `<button type="button" onClick={addMore}>Add</button>`

imonroe commented:
Ok.  I was going off the pattern you used here: https://github.com/SU-SWS/react_paragraphs/blob/60d12ad2afd053d777521fb631c7a59b7aa14549/js/src/Components/Widgets/MediaLibrary.js#L120

But I see you've used both that and the `button` syntax in various places.  I'll update it.

imonroe committed:
D8CORE-2106: Changed the input elements to `button` elements

imonroe commented:
It's now set up to use `button` elements rather than input elements.

imonroe committed:
D8CORE-2106: buttons needed additional class

imonroe committed:
D8CORE-2106: syntax fixup

pookmish commented:
The functionality looks good now. but i'm getting some new error. This occurs after i create a page with a card, populate the link field and save. then when i go back to edit the page, open the modal and then the next page reload will display the error. 
Whats happening that I can tell is that when the modal opens, the react environment calls the rest api endpoint to get the data from the entity with field values. something about the new values are not correct and so it produces an error.

![image](https://user-images.githubusercontent.com/7185045/84198970-3cf93e00-aa59-11ea-99ae-7ec691973b5e.png)

imonroe committed:
D8CORE-2106: attempted bug fix for linkUUID php error

imonroe committed:
Update LinkWidget.js

pookmish commented:
The error was from another patch applied to drupal core. unrelated to this work.
